### PR TITLE
Chitu board support (e.g., Tronxy X5s)

### PR DESCRIPTION
**Pull request by xC0000005**
### Description

This change adds a LD script for the Chitu F103 based boards (and the same encryption scheme is used on pretty much all Chitu boards), and a script to encrypt the resulting firmware so SD based updates can be used.

### Benefits

Allows users to update via SD card, meaning they can revert to Chitu firmware if they so desire. When a PR is submitted to update the SD/LCD pins, this should allow full functionality with the user's choice of firmware.

Note the LD script length doesn't match the offset, but it does match the value the bootloader checks for.

### Related Issues

#7264 
#14655

### CODE REVIEW

File: buildroot/share/PlatformIO/scripts/chitu_crypt.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
+Import("env")
+import struct
+
+# Relocate firmware from 0x08000000 to 0x08007000
```
Comment (tpruvot) [inline buildroot/share/PlatformIO/scripts/chitu_crypt.py]: small typo in comment.. 0x8008800

File: platformio.ini [hunk 2]
```diff
@@ -624,3 +624,19 @@ build_flags = -c -H -std=gnu++11 -Wall -Os -D__MARLIN_FIRMWARE__
...
+
+#
+# Chitu boards like Troxy X5s (STM32F103ZET6)
+#
+[env:chitu_f103]
+platform      = ststm32
+framework     = arduino
+board         = genericSTM32F103ZE
+extra_scripts = buildroot/share/PlatformIO/scripts/chitu_crypt.py
+
```
Comment (tpruvot) [inline platformio.ini]: extra line to remove here, else could be a good practice to place it near the other F1 boards... before mks robin

File: buildroot/share/PlatformIO/scripts/chitu_crypt.py (hunk 1 above)
Comment (xC0000005) [reply buildroot/share/PlatformIO/scripts/chitu_crypt.py]: Fixing now.

File: platformio.ini (hunk 2 above)
Comment (xC0000005) [reply platformio.ini]: Fixed.

Comment (tpruvot) [reply platformio.ini]: nice to see a real/proper use of that... instead of an useless encryption without crc ;)

**Comment by marcio-ao** [comment]
Hummm, is cthulhu support next? :D

File: platformio.ini [hunk 3]
```diff
@@ -528,6 +528,34 @@ lib_ignore    = Adafruit NeoPixel
...
+framework   = arduino
+board       = malyanM200
+build_flags = !python Marlin/src/HAL/HAL_STM32F1/build_flags.py -DMCU_STM32F103CB -D __STM32F1__=1 -std=c++1y -D MOTHERBOARD="BOARD_MALYAN_M200" -DSERIAL_USB -ffunction-sections -fdata-sections -Wl,--gc-sections
+  -DDEBUG_LEVEL=0 -D__MARLIN_FIRMWARE__
+src_filter  = ${common.default_src_filter} +<src/HAL/HAL_STM32F1>
+#-<frameworks>
+lib_ignore  = Adafruit NeoPixel, LiquidCrystal, LiquidTWI2, TMCStepper, U8glib-HAL, SPI
+
+#
+# Chitu boards like Troxy X5s (STM32F103ZET6)
```
Comment (tpruvot) [inline platformio.ini]: Tro**n**XY X5S ?

Comment (xC0000005) [reply platformio.ini]: Fixed.

**Comment by thinkyhead** [comment]
Our current Tronxy X5s configuration assumes an ATmega1280 (Melzi) board. Is this a different model printer or an upgrade board?

**Comment by xC0000005** [comment]
Tronxy sells a bunch of printers with the same name. Your current config is right for some of them, and should probably remain. The config file may be close to right for the printers with the Chitu boards, but I don’t own one, so can’t say for sure.

**Comment by InsanityAutomation** [comment]
I've got the x3s with a chithu to try and test at some point. It'll be awhile though.

**Comment by useless2u** [comment]
Is there any reason for flashing to stop after write addr:801f800 ?

**Comment by xC0000005** [comment]
Yes - if there's no more data to write. The bootloader isn't terribly
helpful - it simply starts the software and doesn't print anything else. If
you see it complete and nothing happens, odds are you don't have your
Serial port set right. What is it set to, and what baud?

**Comment by JHaney** [comment]
How are you guys updating the firmware?
I tried just putting the CBD file onto an SD card root and booting with it but no luck so far though I'm running an x5sa pro board.
I did notice you mentioned small files dont flash

Before starting this I did an STM32FLASH and dumped the firmware out.

Interface serial_posix: 57600 8E1
Version      : 0x22
Option 1     : 0x00
Option 2     : 0x00
Device ID    : 0x0414 (STM32F10xxx High-density)
- RAM        : 64KiB  (512b reserved by bootloader)
- Flash      : 512KiB (size first sector: 2x2048)
- Option RAM : 16b
- System RAM : 2KiB
Memory read
Read address 0x08080000 (100.00%) Done.

[x5sa_pro_stock.zip](https://github.com/MarlinFirmware/Marlin/files/4063538/x5sa_pro_stock.zip)

**Comment by xC0000005** [comment]
I put the SD card in with update.cbd in it. One oddness that's a side
effect of me not understanding *something* (I don't know what) - the
bootloader won't flash the same image twice, and apparently the file key
has something to do with it. I keep a copy of stock firmware and when the
board won't update to mine, I flash to it instead. Here - I extracted the
stock firmware and encrypted it to form a disk update.

**Comment by quarkpower** [comment]
hello and thanks for this proyect

I wanted to ask you if the lcd will be compatible with marlin (doesn't this printer uses different g-m commands?) and if it is already implemented this board support in marlin 2.0

again, thanks a lot (you help a lot of people with this)

**Comment by 1lug1998** [comment]
Hey guys and gals, 

I have a Tronxy XY2 and I have been doing an insane amount of research about updating the Chitu board on it to Marlin 2.0. I am very familiar with Marlin, as I had it on a printer from 5 years ago running on a Mega 2560.

I am having a hard time understanding where exactly this project is at. Is this a supported board now, or not? The board has a STM32F103ZET6 processor. 

If this is a finished project, when modifying the configuration, which motherboard do I define? 
Which pins? 
Do we have the LCD working? 

Thank you

**Comment by thinkyhead** [comment]
@1lug1998 — This Issue Queue is for Marlin bug reports and development-related issues, and we prefer not to handle user-support questions here. (As noted on [this page](https://github.com/MarlinFirmware/Marlin/blob/2.0.x/.github/contributing.md#i-dont-want-to-read-this-whole-thing-i-just-have-a-question).) For best results getting help with specific configurations and troubleshooting, please use the following resources:

- MarlinFW.org [Marlin Documentation](http://marlinfw.org)
- RepRap.org [Marlin Forum](http://forums.reprap.org/list.php?415)
- [Tom's 3D Forums](https://discuss.toms3d.org/)
- Facebook Group ["Marlin Firmware"](https://www.facebook.com/groups/1049718498464482/)
- Facebook Group ["Marlin Firmware for 3D Printers"](https://www.facebook.com/groups/3Dtechtalk/)
- [Marlin Configuration](https://www.youtube.com/results?search_query=marlin+configuration) on YouTube
- Marlin Discord server. Join link: https://discord.gg/n5NJ59y

**Comment by myexec** [comment]
> How are you guys updating the firmware?
> I tried just putting the CBD file onto an SD card root and booting with it but no luck so far though I'm running an x5sa pro board.
> I did notice you mentioned small files dont flash
> 
> Before starting this I did an STM32FLASH and dumped the firmware out.
> 
> Interface serial_posix: 57600 8E1
> Version : 0x22
> Option 1 : 0x00
> Option 2 : 0x00
> Device ID : 0x0414 (STM32F10xxx High-density)
> 
> * RAM        : 64KiB  (512b reserved by bootloader)
> * Flash      : 512KiB (size first sector: 2x2048)
> * Option RAM : 16b
> * System RAM : 2KiB
>   Memory read
>   Read address 0x08080000 (100.00%) Done.
> 
> [x5sa_pro_stock.zip](https://github.com/MarlinFirmware/Marlin/files/4063538/x5sa_pro_stock.zip)

Hi @JHaney,
How did you manage to dump the firmware? What model is your board?
I have CXY-V6-191017 and STM32FLASH could not initalize it trough USB.

**Comment by JHaney** [comment]
@myexec my board is a CXY-V8-190729.
It's been a while but if I remember right the board has to to be in bootloader mode first.
-Put a jumper right of the USB port on the BOOT pins (JP1).
-Power on the board
-Hit the rest button (SW1) near the BOOT jumper which should put your board into bootloader mode.
-The STM32FLASH tool should now be able to talk to it properly.

The mainboard dump this user posted is a gold mine.
https://github.com/MarlinFirmware/Marlin/issues/14655

**Comment by xC0000005** [comment]
That’s right. It’s the boot pin that’s the key.

**Comment by myexec** [comment]
Ok great, I was able to dump the firmware. Now in order to create a backup for install from sd card I need to encrypt it with chitu_crypt.py is that correct ?

> That’s right. It’s the boot pin that’s the key.
> […](#)

**Comment by JHaney** [comment]
The backup pulled from STM32FLASH is already encrypted and can be used to restore the original firmware.
That's about as far as I got attempting to flash marlin as I think the encryption changed on my revision of the board or I simply was doing it wrong (Probably my fault).

To end my tale I went from this to an SKR1.3 with TMC2209's.
FYI G34 Z Steppers Auto Leveling is the cat's pajamas and then some.

Good luck!

**Comment by xC0000005** [comment]
The backup is NOT encrypted, but that’s ok, you only need that terrible encryption if you’re going to use the disk based firmware update. If you want to flash directly on, whatever happens is between you and the hardware. :)

**Comment by myexec** [comment]
Yes it did not seem encrypted to me, but I was trying to make a back up file for disk based firmware update, so the question is if encrypting the dump file with chitu_crypt.py script will work?

> The backup is NOT encrypted, but that’s ok, you only need that terrible encryption if you’re going to use the disk based firmware update. If you want to flash directly on, whatever happens is between you and the hardware. :)
> […](#)

**Comment by xC0000005** [comment]
Oh, ok, to do that, you need to strip out the bootloader (first 0x8800),
AND strip all the 0xFF bytes at the end. Encrypt what's left with the
chitu_crypt and you've got a valid .CBD file. Note that I still don't
understand how the file keys work, so sometimes I have to flash with a
"official" firmware first, and then the one I create. I suspect the
bootloader writes a flag somewhere to indicate something about the flashed
firmware that I need to reset, but I don't know what that is.

**Comment by myexec** [comment]
Ok I confirm the chitu_crypt works for CXY-V6-191017 board, I was able to encrypt the dumped firmware using the method above and installed it back. Thanks @xC0000005 :) There is a slight chance this board won't go to the trash :)
Now I flashed it with Marlin but can't connect via usb. I can see the serial port but there is no response. What settings should I have in configuration.h ?
I tried #define SERIAL_PORT 0 but cant compile. It compiles with -1 and 1 but no response.

**Comment by xC0000005** [comment]
Serial port should be set to 0, don’t bother with USB, this board uses a CH340 serial bridge. Disable SD support, LCD, all that fun stuff and see if it boots up then.

**Comment by rhapsodyv** [comment]
My board is CXY-V5-180409 ... do BOARD_CHITU3D works im my board? anyone already tried?

**Comment by xC0000005** [comment]
If it’s a 103Z, yes, it’s pretty much the same. You’ll need to overwrite the bootloader for now and the TFT doesn’t work yet (I’m looking into it).

**Comment by rhapsodyv** [comment]
Do you have any tutorial about how to backup the current firmware and to burn the new one?

> If it’s a 103Z, yes, it’s pretty much the same. You’ll need to overwrite the bootloader for now and the TFT doesn’t work yet (I’m looking into it).
> […](#)

**Comment by rhapsodyv** [comment]
@xC0000005 
Seems some one got the TFT working: https://www.facebook.com/groups/169728600253557/permalink/679819429244469/

https://drive.google.com/drive/folders/15VEdWYZjFFmqWIbHOoQjOLDZgS4z9PcP?fbclid=IwAR2OR8ia2m2cjOH1Y2JqwSLgEwp3lfOl17pGPXAuwUG_B9P9IFv6hBnZbmU

I'm trying contact to see how he does this...

> If it’s a 103Z, yes, it’s pretty much the same. You’ll need to overwrite the bootloader for now and the TFT doesn’t work yet (I’m looking into it).
> […](#)

**Comment by xC0000005** [comment]
Sounds great to me. If everything works, all I have to do is get SD update working with full marlin and people won’t need to use the CUBE programmer/pull the boot pins, etc.

**Comment by MrPoke21** [comment]
> 
> 
> @xC0000005
> Seems some one got the TFT working: https://www.facebook.com/groups/169728600253557/permalink/679819429244469/
> 
> https://drive.google.com/drive/folders/15VEdWYZjFFmqWIbHOoQjOLDZgS4z9PcP?fbclid=IwAR2OR8ia2m2cjOH1Y2JqwSLgEwp3lfOl17pGPXAuwUG_B9P9IFv6hBnZbmU
> 
> I'm trying contact to see how he does this...
> 
> > If it’s a 103Z, yes, it’s pretty much the same. You’ll need to overwrite the bootloader for now and the TFT doesn’t work yet (I’m looking into it).
> > […](#)

https://github.com/MrPoke21/Marlin-Chitu-board-support   ;)

**Pull request merged**